aten/src/ATen/native/cpu/Reduce.h

```diff
@@ -118,9 +118,10 @@ static inline void vectorized_outer_reduction(char** data, int64_t inner_stride,
 template<typename traits, typename res_t>
 static void set_result(const int index, const res_t result, const TensorIteratorBase &iter, const int num_outputs) {
   // static_assert(std::is_same<res_t, typename traits::arg2_t>::value, "data types must match");
+  using out_t = typename traits::arg2_t;
```

Comment:
Test failure is real, this breaks `argmax` and `argmin` where the result is not either of the input types. I think `project` really does need to be updated in `SharedReduceOps.h`.

Reply:
Oh so the issue here is that cpu reduce kernel bytecasts the results in the hopes that width is the same and it'll somehow work out (pre-this-pr), and long time ago it had an assert to make sure that it at least makes some sense, but abandoned it long ago? I'd still say that's a very error-prone cpu kernel, and problems should be fixed here.

Reply:
Yes, I need to rewrite the code as GPU path to fix it.

Reply:
> Oh so the issue here is that cpu reduce kernel bytecasts the results in the hopes that width is the same and it'll somehow work out (pre-this-pr)

No, I don't think there's any byte-casting involved. Conceptually, the reduction involves three types: 
- `data_t` - the input type
- `acc_t` - the accumulator type
- `res_t` - the result type

and the type signatures of the reduction op should be:
```
reduce(acc_t, data_t) -> acc_t
combine(acc_t, acc_t) -> acc_t
project(acc_t) -> res_t
```
(This is slightly simplified)

Some of the norm operators have `project(acc_t) -> acc_t` which assumes that `acc_t` is the same as `res_t`. This PR breaks that assumption by accumulating in a higher precision type.

The workaround proposed by this PR is to cast the output of `project` to `data_t`, which forces the output and input dtypes to match. This need not be true in general and in particular is not true for `argmin` and `argmax`.

The correct fix IMO is to change `project` to the correct output dtype in `SharedReduceOps.h` and not touch `Reduce.h`, other than maybe removing the assert that was commented out.

Reply:
Thanks, but this **project(acc_t) -> out_t** is also used by GPU reduce kernel, I need to check it whether is ok just change it.